File: libs/cubeinterface/MultiParameterHypothesis.cc
#include "MultiParameterHypothesis.h"
#include "Utilities.h"
#include <cmath>
#include <sstream>
#include <limits>
#include <cassert>

namespace EXTRAP
{
MultiParameterHypothesis::MultiParameterHypothesis( void ) : m_function( NULL ), m_function_refcount( NULL )
{
}

MultiParameterHypothesis::MultiParameterHypothesis( MultiParameterFunction* f )
{
    this->m_function                 = f;
    this->m_function_refcount        = new HypothesisFunctionRefcount();
    this->m_function_refcount->count = 1; // Initially referenced only inside this hypothesis
    this->m_function_refcount->func  = f;
    DebugStream << "### Initialized refcount to 1 for " << this->m_function << " at " << this->m_function_refcount << "\n";
}

MultiParameterFunction*
MultiParameterHypothesis::getFunction( void ) const
{
    return this->m_function;
}

double
MultiParameterHypothesis::getRSS( void ) const
{
    return this->m_RSS;
}

void
MultiParameterHypothesis::setRSS( double rss )
{
    this->m_RSS = rss;
}

double
MultiParameterHypothesis::getrRSS( void ) const
{
    return this->m_rRSS;
}

void
MultiParameterHypothesis::setrRSS( double rrss )
{
    this->m_rRSS = rrss;
}


double
MultiParameterHypothesis::getAR2( void ) const
{
    return this->m_AR2;
}

void
MultiParameterHypothesis::setAR2( double ar2 )
{
    this->m_AR2 = ar2;
}

double
MultiParameterHypothesis::getSMAPE( void ) const
{
    return this->m_SMAPE;
}

void
MultiParameterHypothesis::setSMAPE( double smape )
{
    this->m_SMAPE = smape;
}

double
MultiParameterHypothesis::getRE( void ) const
{
    return this->m_RE;
}

void
MultiParameterHypothesis::setRE( double re )
{
    this->m_RE = re;
}

std::vector<double>
MultiParameterHypothesis::getPredictedPoints( void )
{
    return this->m_predicted_points;
}

std::vector<double>
MultiParameterHypothesis::getActualPoints( void )
{
    return this->m_actual_points;
}

std::vector<double>
MultiParameterHypothesis::getPs( void )
{
    return this->m_ps;
}

std::vector<double>
MultiParameterHypothesis::getSizes( void )
{
    return this->m_sizes;
}

void
MultiParameterHypothesis::computeCost( const std::vector<DataPoint>& points )
{
    this->m_RSS  = 0;
    this->m_rRSS = 0;
    double smape  = 0;
    double re_sum = 0;
    std::vector<double> predicted_points;
    std::vector<double> actual_points;
    std::vector<double> ps;
    std::vector<double> sizes;

    for ( int i = 0; i < points.size(); i++ )
    {
        double predicted           = this->m_function->evaluate( points[ i ].getParameterList() );

        const ParameterValueList& list = points[ i ].getParameterList();
        Parameter p = Parameter("p");
        Parameter s = Parameter("size");
        ParameterValueList::const_iterator it = list.find(p);
        double processors = it->second;
        //std::cout << "Processors: " << processors << "\n";
        ParameterValueList::const_iterator it2 = list.find(s);
        double size = it2->second;
        //std::cout << "Size: " << size << "\n";


        predicted_points.push_back(predicted);
        double actual              = points[ i ].getValue();

        //std::cout << "predicted: " << predicted << "\n";
        //std::cout << "actual: " << actual << "\n";
        ps.push_back(processors);
        sizes.push_back(size);

        actual_points.push_back(actual);
        double difference          = predicted - actual;
        double absolute_difference = fabs( difference );
        double abssum              = fabs( actual ) + fabs( predicted );

        // calculate relative error
        double absolute_error = fabs( predicted - actual );
        //double tmp = std::max(absolute_error, actual);
        double relative_error = absolute_error / actual;
        re_sum = re_sum + relative_error;

        this->m_RSS += difference * difference;
        double relativeDifference = difference / points[ i ].getValue();
        this->m_rRSS += relativeDifference * relativeDifference;

        if ( abssum != 0.0 )
        {
            // This `if` condition prevents a division by zero, but it is correct: if sum is 0, both `actual` and `predicted`
            // must have been 0, and in that case the error at this point is 0, so we don't need to add anything.
            smape += fabs( difference ) / abssum * 2;
        }
    }
    //times 100 for percentage error
    this->m_RE    = re_sum / points.size();
    this->m_SMAPE = smape / points.size() * 100;
    this->m_predicted_points = predicted_points;
    this->m_actual_points = actual_points;
    this->m_ps = ps;
    this->m_sizes = sizes;
}

void
MultiParameterHypothesis::computeAdjustedRSquared( double TSS, const std::vector<DataPoint>& points )
{
    this->m_AR2 = 0.0;
    double adjR = 1.0 - ( this->m_RSS / TSS );

    double counter = 0;
    for ( int i = 0; i < this->m_function->getMultiParameterTerms().size(); i++ )
    {
        counter += m_function->getMultiParameterTerms()[ i ].getCompoundTermParameterPairs().size();
    }

    double degrees_freedom = points.size() - counter - 1;

    this->m_AR2 = ( 1.0 - ( 1.0 - adjR ) * ( points.size() - 1.0 ) / degrees_freedom );

    return;
}

void
MultiParameterHypothesis::computeCostLeaveOneOutCrossValidation( const std::vector<DataPoint>& points, DataPoint missing )
{
    double predicted  = this->m_function->evaluate( missing.getParameterList() );
    double actual     = missing.getValue();
    double difference = predicted - actual;
    this->m_RSS += difference * difference;
    double relativeDifference = difference / missing.getValue();
    this->m_rRSS += relativeDifference * relativeDifference;
    double abssum = fabs( actual ) + fabs( predicted );
    if ( abssum != 0.0 )
    {
        this->m_SMAPE += ( fabs( difference ) / abssum * 2 ) / points.size() * 100;
    }
}

void
MultiParameterHypothesis::estimateParameters( const std::vector<DataPoint>& points )
{
    assert( m_function != NULL );

    int hypothesisTotalTerms = this->m_function->getMultiParameterTerms().size() + 1;
    int experimentCount      = points.size();

    double*             M              = new double[ hypothesisTotalTerms * hypothesisTotalTerms ];
    double*             f              = new double[ hypothesisTotalTerms ];
    double*             rhs            = new double[ experimentCount ];
    double*             regularization = new double[ hypothesisTotalTerms ];
    std::vector<double> AT;
    AT.resize( hypothesisTotalTerms * experimentCount );

    for ( int i = 0; i < AT.size(); i++ )
    {
        AT[ i ] = 0;
    }
    for ( int i = 0; i < hypothesisTotalTerms; i++ )
    {
        int index = i * experimentCount;
        for ( int rhs_index = 0; rhs_index < experimentCount; rhs_index++ )
        {
            AT[ index ] = 1;

            double eval;
            if ( i > 0 )
            {
                MultiParameterTerm& multiParameterTerm = ( this->m_function->getMultiParameterTerms() )[ i - 1 ];

                multiParameterTerm.setCoefficient( 1 ); // reset coefficient

                for ( int j = 0; j < multiParameterTerm.getCompoundTermParameterPairs().size(); j++ )
                {
                    CompoundTerm t = ( multiParameterTerm.getCompoundTermParameterPairs() )[ j ].first;
                    t.setCoefficient( 1.0 );
                }

                eval = multiParameterTerm.evaluate( points[ rhs_index ].getParameterList() );
            }
            else
            {
                eval = 1;
            }
            AT[ index ] *= eval;
            index++;
            rhs[ rhs_index ] = points[ rhs_index ].getValue();
        }
    }

    for ( int i = 0; i < hypothesisTotalTerms; i++ )
    {
        double reg = 1;
        int    tmp = floor( log10( AT[ experimentCount * i + experimentCount - 1 ] ) );
        if ( tmp > 0 )
        {
            reg = pow( 10.0f, tmp );
        }
        regularization[ i ] = reg;
        for ( int j = 0; j < experimentCount; j++ )
        {
            int index = experimentCount * i + j;
            AT[ index ] *= reg;
        }
    }

    for ( int i = 0; i < hypothesisTotalTerms; i++ )
    {
        int row = i * experimentCount;
        for ( int j = 0; j < hypothesisTotalTerms; j++ )
        {
            int t_row = experimentCount * j;
            int index = i * hypothesisTotalTerms + j;
            M[ index ] = 0;
            for ( int k = 0; k < experimentCount; k++ )
            {
                M[ index ] += AT[ row + k ] * AT[ t_row + k ];
            }
        }
    }
    for ( int i = 0; i < hypothesisTotalTerms; i++ )
    {
        int row = i * experimentCount;
        f[ i ] = 0;
        for ( int k = 0; k < experimentCount; k++ )
        {
            f[ i ] += AT[ row + k ] * rhs[ k ];
        }
    }

    matrixInverse( M, hypothesisTotalTerms );

    for ( int i = 0; i < hypothesisTotalTerms; i++ )
    {
        int irow = i * hypothesisTotalTerms;
        if ( i == 0 )
        {
            this->m_function->setConstantCoefficient( 0 );
        }
        else
        {
            ( this->m_function->getMultiParameterTerms() )[ i - 1 ].setCoefficient( 0 );
        }

        for ( int j = 0; j < hypothesisTotalTerms; j++ )
        {
            if ( i == 0 )
            {
                this->m_function->setConstantCoefficient( this->m_function->getConstantCoefficient() +  ( M[ irow + j ] * f[ j ] ) );
            }
            else
            {
                ( this->m_function->getMultiParameterTerms() )[ i - 1 ].setCoefficient( ( this->m_function->getMultiParameterTerms() )[ i - 1 ].getCoefficient() + ( M[ irow + j ] * f[ j ] ) );
            }
        }
        if ( i == 0 )
        {
            this->m_function->setConstantCoefficient( this->m_function->getConstantCoefficient() * regularization[ i ] );
        }
        else
        {
            ( this->m_function->getMultiParameterTerms() )[ i - 1 ].setCoefficient( ( this->m_function->getMultiParameterTerms() )[ i - 1 ].getCoefficient() * regularization[ i ] );
        }
    }

    delete[] M;
    delete[] f;
    delete[] rhs;
    delete[] regularization;
    AT.clear();
    return;
}

bool
MultiParameterHypothesis::isValid( void )
{
    double cost = m_RSS;
    return !( cost != cost || fabs( cost ) == std::numeric_limits<double>::infinity() );
}

double
MultiParameterHypothesis::calculateMaximalTermContribution( int termIndex, const std::vector<DataPoint>& points ) const
{
    std::vector<EXTRAP::MultiParameterTerm> mt = m_function->getMultiParameterTerms();
    assert( termIndex < mt.size() );
    double max = 0;
    for ( int i = 0; i < points.size(); i++ )
    {
        double contribution = fabs( mt[ termIndex ].evaluate( points[ i ].getParameterList() ) / points[ i ].getValue() );
        if ( contribution > max )
        {
            max = contribution;
        }
    }
    return max;
}

void
MultiParameterHypothesis::cleanConstantCoefficient( double epsilon, const std::vector<DataPoint>& points )
{
    // This function is used to correct numerical imprecision in the matrix caculations,
    // when the constant coefficient should be zero but is instead very small.
    // We take into account the minimum data value to make sure that we don't "nullify"
    // actually relevant numbers.
    std::vector<DataPoint>::const_iterator it      = points.begin();
    double                                 minimum = it->getValue();
    ++it;
    for (; it != points.end(); ++it )
    {
        if ( it->getValue() < minimum )
        {
            minimum = it->getValue();
        }
    }

    if ( fabs( this->m_function->getConstantCoefficient() / minimum ) < epsilon )
    {
        this->m_function->setConstantCoefficient( 0 );
    }
}

void
MultiParameterHypothesis::keepFunctionAlive( void )
{
    assert( this->m_function_refcount != NULL );
    assert( this->m_function_refcount->func == this->m_function );
    this->m_function_refcount->count++;

    //DebugStream << "Incremented external function refcount (" << this->m_function << ") at " << this->m_function_refcount << " to " << this->m_function_refcount->count << "\n";
}

void
MultiParameterHypothesis::freeFunction( void )
{
    assert( this->m_function_refcount != NULL );
    assert( this->m_function_refcount->func == this->m_function );
    ( this->m_function_refcount->count )--;
    //DebugStream << "Decremented external function refcount (" << this->m_function << ") at " << this->m_function_refcount << " to " << this->m_function_refcount->count << "\n";
    if ( this->m_function_refcount->count == 0 )
    {
        DebugStream << "### Deleting function (" << this->m_function << ") with refcount at " << this->m_function_refcount << "\n";
        delete this->m_function;
        delete this->m_function_refcount;
        this->m_function          = NULL;
        this->m_function_refcount = NULL;
    }
}
};

File: src/entities/multi_parameter_hypothesis.py
"""
This file is part of the Extra-P software (http://www.scalasca.org/software/extra-p)

Copyright (c) 2020,
Technische Universitaet Darmstadt, Germany
 
This software may be modified and distributed under the terms of
a BSD-style license.  See the COPYING file in the package base
directory for details.
"""


import numpy

    
class MultiParameterHypothesis:
    """
    This class represents a multi parameter hypothesis, it is used to represent
    a performance function with several parameters. However, it can have also
    only one parameter. The modeler calls many of these objects to find the best
    model that fits the data.
    """
    
    
    def __init__(self, function, median):
        """
        Initialize MultiParameterHypothesis object.
        """
        self.function = function
        self.RSS = 0
        self.rRSS = 0
        self.SMAPE = 0
        self.AR2 = 0
        self.RE = 0
        self.median = median


    def get_function(self):
        """
        Return the function.
        """
        return self.function
    
    
    def get_RSS(self):
        """
        Return the RSS.
        """
        return self.RSS


    def get_rRSS(self):
        """
        Return the rRSS.
        """
        return self.rRSS


    def get_AR2(self):
        """
        Return the AR2.
        """
        return self.AR2


    def get_SMAPE(self):
        """
        Return the SMAPE.
        """
        return self.SMAPE
    
    
    def get_RE(self):
        """
        Return the RE.
        """
        return self.RE
    
    
    def set_RSS(self, RSS):
        """
        Set the RSS.
        """
        self.RSS = RSS


    def set_rRSS(self, rRSS):
        """
        Set the rRSS.
        """
        self.rRSS = rRSS


    def set_AR2(self, AR2):
        """
        Set the AR2.
        """
        self.AR2 = AR2


    def set_SMAPE(self, SMAPE):
        """
        Set the SMAPE.
        """
        self.SMAPE = SMAPE
    
    
    def set_RE(self, RE):
        """
        Set the RE.
        """
        self.RE = RE
    
    
    def clean_constant_coefficient(self, phi, measurements):
        """
        This function is used to correct numerical imprecision in the caculations,
        when the constant coefficient should be zero but is instead very small.
        We take into account the minimum data value to make sure that we don't "nullify"
        actually relevant numbers.
        """
        minimum = 0
        for measurements_id in range(len(measurements)):
            if measurements_id == 0:
                if self.median == True:
                    minimum = measurements[measurements_id].get_value_median()
                else:
                    minimum = measurements[measurements_id].get_value_mean()
            else:
                if self.median == True:
                    if measurements[measurements_id].get_value_median() < minimum:
                        minimum = measurements[measurements_id].get_value_median()
                else:
                    if measurements[measurements_id].get_value_mean() < minimum:
                        minimum = measurements[measurements_id].get_value_mean()

        if abs(self.function.get_constant_coefficient() / minimum) < phi:
            self.function.set_constant_coefficient(0)
    

    def compute_cost(self, measurements, coordinates):
        """
        Compute the cost for a multi parameter hypothesis.
        """
        self.RSS = 0
        self.rRSS = 0
        smape = 0
        re_sum = 0
        
        for i in range(len(measurements)):
            measurement = measurements[i]
            coordinate_id = measurement.get_coordinate_id()
            coordinate = coordinates[coordinate_id]
            dimensions = coordinate.get_dimensions()
            parameter_value_pairs = {}
            for i in range(dimensions):
                parameter, value = coordinate.get_parameter_value(i)
                parameter_value_pairs[parameter.get_name()] = float(value)
            
            predicted = self.function.evaluate(parameter_value_pairs)
            #print(predicted)
            if self.median == True:
                actual = measurement.get_value_median()
            else:
                actual = measurement.get_value_mean()
            #print(actual)
            
            difference = predicted - actual
            #absolute_difference = abs(difference)
            abssum = abs(actual) + abs(predicted)

            # calculate relative error
            absolute_error = abs(predicted - actual)
            relative_error = absolute_error / actual
            re_sum = re_sum + relative_error

            self.RSS += difference * difference
            if self.median == True:
                relativeDifference = difference / measurement.get_value_median()
            else:
                relativeDifference = difference / measurement.get_value_mean()
            self.rRSS += relativeDifference * relativeDifference
    
            if abssum != 0.0:
                # This `if` condition prevents a division by zero, but it is correct: if sum is 0, both `actual` and `predicted`
                # must have been 0, and in that case the error at this point is 0, so we don't need to add anything.
                smape += abs(difference) / abssum * 2
      
        # times 100 for percentage error
        self.RE    = re_sum / len(measurements)
        self.SMAPE = smape / len(measurements) * 100
        

    def compute_adjusted_rsquared(self, TSS, measurements):
        """
        Compute the adjusted R^2 for the hypothesis.
        """
        self.AR2 = 0.0
        adjR = 1.0 - (self.RSS / TSS)
        counter = 0
        
        for i in range(len(self.function.get_multi_parameter_terms())):
            counter += len(self.function.get_multi_parameter_terms()[i].get_compound_term_parameter_pairs())
        
        degrees_freedom = len(measurements) - counter - 1
        self.AR2 = ( 1.0 - (1.0 - adjR) * (len(measurements) - 1.0) / degrees_freedom )
  
  
    def compute_coefficients(self, measurements, coordinates):
        """
        Computes the coefficients of the function using the least squares solution.
        """
        hypothesis_total_terms = len(self.function.get_multi_parameter_terms()) + 1
        
        # creating a numpy matrix representation of the lgs
        a_list = []
        b_list = []
        for element_id in range(len(measurements)):
            if self.median == True:
                value = measurements[element_id].get_value_median()
            else:
                value = measurements[element_id].get_value_mean()
            list_element = []
            for multi_parameter_term_id in range(hypothesis_total_terms):
                if multi_parameter_term_id == 0:
                    list_element.append(1)
                else:
                    multi_parameter_term = self.function.get_multi_parameter_term(multi_parameter_term_id-1)
                    #_, parameter_value = coordinates[element_id].get_parameter_value(0)
                    coordinate_id = measurements[element_id].get_coordinate_id()
                    coordinate = coordinates[coordinate_id]
                    dimensions = coordinate.get_dimensions()
                    parameter_value_pairs = {}
                    for i in range(dimensions):
                        parameter, value = coordinate.get_parameter_value(i)
                        parameter_value_pairs[parameter.get_name()] = float(value)
                    multi_parameter_term_value = multi_parameter_term.evaluate(parameter_value_pairs)
                    list_element.append(multi_parameter_term_value)
            a_list.append(list_element)
            b_list.append(value)
            #print(str(list_element)+"[x]=["+str(value)+"]")
            #logging.debug(str(list_element)+"[x]=["+str(value)+"]")
            
        # solving the lgs for X to get the coefficients
        A = numpy.array(a_list)
        B = numpy.array(b_list)
        X = numpy.linalg.lstsq(A,B,None)
        #print("Coefficients:"+str(X[0]))
        #logging.debug("Coefficients:"+str(X[0]))
        
        # setting the coefficients for the hypothesis
        self.function.set_constant_coefficient(X[0][0])
        for multi_parameter_term_id in range(hypothesis_total_terms-1):
            self.function.get_multi_parameter_term(multi_parameter_term_id).set_coefficient(X[0][multi_parameter_term_id+1])
            
            
    def is_valid(self):
        """
        Checks if there is a numeric imprecision. If this is the case the hypothesis will be ignored.
        """
        valid = not (self.RSS != self.RSS or abs(self.RSS) == float('inf'))
        return valid
        
        
    def calc_term_contribution(self, term_index, measurements, coordinates):
        """
        Calculates the term contribution of the term with the given term id to see if it is smaller than epsilon.
        """
        multi_parameter_terms = self.function.get_multi_parameter_terms()
        #compound_terms = self.function.get_compound_terms()
        maximum_term_contribution = 0
        for element_id in range(len(measurements)):
            #_, parameter_value = coordinates[element_id].get_parameter_value(0)
            dimensions = coordinates[element_id].get_dimensions()
            parameter_value_pairs = {}
            for i in range(dimensions):
                parameter, value = coordinates[element_id].get_parameter_value(i)
                parameter_value_pairs[parameter.get_name()] = float(value)
            if self.median == True:
                contribution = abs( multi_parameter_terms[term_index].evaluate(parameter_value_pairs) / measurements[element_id].get_value_median())
            else:
                contribution = abs( multi_parameter_terms[term_index].evaluate(parameter_value_pairs) / measurements[element_id].get_value_mean())
            if contribution > maximum_term_contribution:
                maximum_term_contribution = contribution
        return maximum_term_contribution
    
    